File: src/features/typing/data/go/diameter_of_binary_tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func diameterOfBinaryTree(root *TreeNode) int {
	best := 0
	var depth func(node *TreeNode) int
	depth = func(node *TreeNode) int {
		if node == nil {
			return 0
		}
		l := depth(node.Left)
		r := depth(node.Right)
		if l+r > best {
			best = l + r
		}
		return max(l, r) + 1
	}
	depth(root)
	return best
}

File: src/features/typing/data/ts/diameter_of_binary_tree.ts
class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

function diameterOfBinaryTree(root: TreeNode | null): number {
    let best = 0;
    const depth = (node: TreeNode | null): number => {
        if (node === null) {
            return 0;
        }
        const l = depth(node.left);
        const r = depth(node.right);
        best = Math.max(best, l + r);
        return Math.max(l, r) + 1;
    };
    depth(root);
    return best;
}
